metadata/src/main/java/org/apache/kafka/controller/LocalLogManager.java

```diff
@@ -0,0 +1,801 @@
+                LeaderInfo claim = null;
+                while (true) {
+                    List<ApiMessageAndVersion> toWrite = null;
+                    long shouldRenounceEpoch = -1, shouldClaimEpoch = -1;
+                    ScribeState curState;
+                    lock.lock();
+                    try {
+                        if (shuttingDown) {
+                            break;
+                        }
+                        if (shouldLead) {
+                            // If shouldLead is true then we need to become a leader.
+                            // This requires reading all the previous log entries, and
+                            // writing out our claim to the log file.
+                            if (state != ScribeState.LEADER) {
+                                // If claim is non-null here, that means we are ready
+                                // to become a leader.  Otherwise, we have to enter
+                                // the BECOME_LEADER state.
+                                if (claim != null) {
+                                    state = ScribeState.LEADER;
+                                    incomingWrites = null;
+                                    leaderInfo = claim;
+                                    index = 0;
+                                    nextWriteIndex = 0;
+                                    shouldClaimEpoch = leaderInfo.epoch;
+                                    claimedCond.signalAll();
+                                } else {
+                                    state = ScribeState.BECOMING_LEADER;
+                                }
+                            }
+                        } else if (state != ScribeState.FOLLOWER) {
+                            // If shouldLead is false, and we're not already a follower,
+                            // become one immediately.
+                            shouldRenounceEpoch = leaderInfo.epoch;
+                            state = ScribeState.FOLLOWER;
+                            nextLogCheckMs = 0;
+                            renouncedCond.signalAll();
+                        }
+                        if (shouldClaimEpoch == -1 && shouldRenounceEpoch == -1) {
+                            switch (state) {
+                                case FOLLOWER:
+                                    // Followers wait until they get woken up, or until a
+                                    // few milliseconds have elapsed, before rechecking
+                                    // the log.
+                                    long now = SystemTime.SYSTEM.milliseconds();
+                                    if (nextLogCheckMs > now) {
+                                        wakeCond.await(nextLogCheckMs - now,
+                                            TimeUnit.MILLISECONDS);
+                                        now = SystemTime.SYSTEM.milliseconds();
+                                    }
```

Comment:
Might be good to append to the comment stating that the spotbugs warning about ignoring the return value is suppressed because we don't care about timing out.

Reply:
This is one of the areas where I disagree with spotbugs.  I actually don't think ignoring the return value of `await` is unusual at all.  Traditionally, you can always get spurious wakeups with condition variables.  Checking the Javadoc for Condition.java reveals that this is the case for Java's condition variables as well.  So what can you conclude from the fact that the timeout has elapsed or not elapsed?  Pretty much nothing, since maybe you just got a bogus wakeup and the condition you are checking for is still not true.

It was actually probably an API design mistake for them to even return a boolean from this method, since it encourages people to write code assuming the condition was triggered if await returns false.  It isn't necessarily so.